File: 486. Predict the Winner.py
# link - https://leetcode.com/problems/predict-the-winner/

class Solution:
    def PredictTheWinner(self, nums: List[int]) -> bool:
        
        def helper(start, end):
            if start == end:
                return nums[start]

            ans = max(
                nums[start] - helper(start+1, end),
                nums[end] - helper(start, end-1)
            )

            return ans
        return helper(0, len(nums)-1) >= 0

File: Leetcode/486. Predict the Winner.cpp
/**
 * https://leetcode.com/problems/predict-the-winner/
 * Time: O(N^2)
 * Space: O(N)
 */

class Solution {
public:
    bool PredictTheWinner(vector<int>& nums) {
        
        int n = nums.size();
        vector<int> dp(n);
        
        for (int i = n - 1; i >= 0; --i){
            for (int j = i; j < n; ++j){
                if (i == j)
                    dp[j] = nums[i];
                else
                    dp[j] = max(nums[i] - dp[j], nums[j] - dp[j - 1]);
            }
        }
        
        return dp[n - 1] >= 0;
    }
};